luarules/gadgets/unit_weapon_smart_select_helper.lua

```diff
@@ -0,0 +1,217 @@
+3. in the unit's .bos animation script, #include "smart_weapon_select.h"  ideally at the beginning of the file.
+4. in the preferred AimWeaponX() function, add the following at the beginning:
+	if (AimingState != AIMING_PRIORITY){
+		return(0);
+	}
+5. in the deferred AimWeaponX() function, add the following at the beginning:
+	if (AimingState != AIMING_BACKUP){
+		return(0);
+	}
+6. If using a dummy weapon, return (0); in its AimWeaponX() function and QueryWeaponX(piecenum) should be set to static piece lower than the turret.
+
+****OPTIONAL*****
+use weapondef.customparams.smart_error_frames to override the default reloadtime derivitive frames error threshold
+]]
+
+--static
+local frameCheckModulo = Game.gameSpeed
+local failedToFireMultiplier = Game.gameSpeed * 1.25
+local minimumFailedToFireFrames = Game.gameSpeed * 4
+local aggroDecayRate = 0.65 --aggro is multiplied by this until it falls within priority aiming state range
+local aggroDecayCap = 10 -- this caps the aggro decay so that error state can last a significant amount of time
+local errorTallyDecayRate = 0.99  -- the error penalty that makes the error state last longer the more cumulative times it happens decays at this rate
+local errorMultiplierAddition = 1 -- every time the error happens, it increases by this number. The tally is squared to make each cumulative error exponentially more punishing
+local PRIORITY_AIMINGSTATE = 1
+local BACKUP_AIMINGSTATE = 2
+
+local priorityAutoAggro = 12 -- how much aggro is accumulated per frameCheckModulo that the priority weapon successfully aims
+local priorityManualAggro = priorityAutoAggro * 4 -- how much aggro is accumulated per frameCheckModulo that the priority weapon successfully aims with a manually assigned target
+local prioritySwitchThreshold = -1  --the aggro at which priority weapon switch is triggered. Aggro is decayed closer to 0 every frameCheckModulo
+
+local backupAutoAggro = 4 -- how much aggro is accumulated per frameCheckModulo that the priority weapon fails to aim
+local backupManualAggro = priorityAutoAggro * 3 --how much aggro is accumulated per frameCheckModulo that the priority weapon fails to aim with a manually assigned target
+local backupErrorAggro = -300 --how much aggro is given multiplied by the errorTallyMultiplier^2 when priority weapon fails to fire.
+local backupSwitchThreshold = -backupAutoAggro * 1.5 --the aggro at which backup weapon switch is triggered. Aggro is decayed closer to 0 every frameCheckModulo
+
+--variables
+local gameFrame = 0
+
+--functions
+local spCallCOBScript = Spring.CallCOBScript
+local spGetUnitWeaponHaveFreeLineOfFire = Spring.GetUnitWeaponHaveFreeLineOfFire
+local spGetUnitWeaponTarget = Spring.GetUnitWeaponTarget
+local spGetUnitWeaponState = Spring.GetUnitWeaponState
+local spGetUnitEstimatedPath = Spring.GetUnitEstimatedPath
+local mathMin = math.min
+local mathMax = math.max
+
+--tables
+local smartUnits = {}
+local unitDefData = {}
```

Comment:
```suggestion
local unitSmartWeapons = {}
```
More descriptive name

Reply:
i like, done